src/llvm-multiversioning.cpp

```diff
@@ -0,0 +1,1067 @@
+            assert(!val->isConstant());
+            auto fid = get_func_id(orig_f);
+            auto addr = ConstantExpr::getPtrToInt(val, T_size);
+            if (info.offset)
+                addr = ConstantExpr::getAdd(addr, ConstantInt::get(T_size, info.offset));
+            gv_relocs.emplace_back(addr, fid);
+            val->setInitializer(rewrite_gv_init(stack));
+        }
+        return changed;
+    };
+    for (auto orig_f: orig_funcs) {
+        if (!has_cloneall && !cloned.count(orig_f))
+            continue;
+        while (single_pass(orig_f)) {
+        }
+    }
+}
+
+std::pair<uint32_t,GlobalVariable*> CloneCtx::get_reloc_slot(Function *F)
+{
+    // Null initialize so that LLVM put it in the correct section.
+    auto id = get_func_id(F);
+    auto &slot = const_relocs[id];
+    if (!slot)
+        slot = new GlobalVariable(M, T_pvoidfunc, false, GlobalVariable::InternalLinkage,
+                                  ConstantPointerNull::get(T_pvoidfunc),
+                                  F->getName() + ".reloc_slot");
+    return std::make_pair(id, slot);
+}
+
+template<typename Stack>
+Value *CloneCtx::rewrite_inst_use(const Stack& stack, Value *replace, Instruction *insert_before)
+{
+    SmallVector<Constant*, 8> args;
+    uint32_t nlevel = stack.size();
+    for (uint32_t i = 1; i < nlevel; i++) {
+        auto &frame = stack[i];
+        auto val = frame.val;
+        Use *use = frame.use;
+        unsigned idx = use->getOperandNo();
+        if (auto expr = dyn_cast<ConstantExpr>(val)) {
+            auto inst = expr->getAsInstruction();
+            inst->replaceUsesOfWith(val->getOperand(idx), replace);
+            inst->insertBefore(insert_before);
+            replace = inst;
+            continue;
+        }
+        unsigned nargs = val->getNumOperands();
+        args.resize(nargs);
+        for (unsigned j = 0; j < nargs; j++) {
+            auto op = val->getOperand(j);
+            if (idx == j) {
+                args[j] = UndefValue::get(op->getType());
+            }
+            else {
+                args[j] = cast<Constant>(op);
+            }
+        }
+        if (auto ary = dyn_cast<ConstantArray>(val)) {
+            replace = InsertValueInst::Create(ConstantArray::get(ary->getType(), args),
+                                              replace, {idx}, "", insert_before);
+        }
+        else if (auto strct = dyn_cast<ConstantStruct>(val)) {
+            replace = InsertValueInst::Create(ConstantStruct::get(strct->getType(), args),
+                                              replace, {idx}, "", insert_before);
+        }
+        else if (isa<ConstantVector>(val)) {
+            replace = InsertElementInst::Create(ConstantVector::get(args), replace,
+                                                ConstantInt::get(T_size, idx), "",
+                                                insert_before);
+        }
+        else {
+            jl_safe_printf("Unknown const use.");
```

Comment:
Also could be a `GlobalIndirectSymbol`. Is the assert actually necessary?

Reply:
The assert is necessary since all of the uses needs to be fixed individually. (AFAICT LLVM doesn't have anything that does this recursive replace...).

This should handle everything that we emit and I don't think LLVM introduces any IFuncs or Aliases for functions automatically.

Reply:
We emit Aliases for functions, Why do we need to fix all uses?

Reply:
> We emit Aliases for functions

For `jl_extern_c` I guess? That's not fixable then.....

> Why do we need to fix all uses?

It's very bad to let the wrong version running.

Reply:
(`GlobalAlias` isn't the not-fixable part. `ExternalLinkage` is.)

Reply:
> It's very bad to let the wrong version running.
  
worse than calling `abort` (instead of, say, a warning)?

Reply:
Well, it may crash or give wrong result and even if that doesn't happen, I do think a compile time error is better than a mysterious runtime performance issue.